buildPythonPackage: add support for conda binary packages

DavHau opened the pull request:
###### Motivation
I have been using these hooks in mach-nix for a while. Conda in general provides quite well defined metadata containing complete dependency trees which makes it often easier to assemble nix expressions for those packages compared to manylinux wheels from pypi.
I think it would be good to add these to nixpkgs as this might open the door for other conda2nix tools as well.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

DavHau committed:
buildPythonPackage: add support for conda

FRidh commented:
It is my intention to get rid of the `format = ...` in time. Instead, users should select the hooks they need based on the build system of the package. Therefore, I think here the changes to `buildPythonPackage` should be undone. Instead, if you want to work with conda, add those hooks to `nativeBuildInputs`.

We also need to agree on who will maintain additions like these in Nixpkgs.

DavHau commented:
OK, I can remove the format and only include the hooks. But then also I'd need to make a package for the conda buildInputs equivalent to the manylinuxXPackages.

In general I'm not a fan of shifting more work to the user. I also think that installing manylinux wheels is still too complicated.
I even would have proposed to create a format `manylinux` that has everything on board.

Why exactly do you want to get rid of the formats? If your motivation is to reduce complexity in `mk-python-derivation.nix`, we could offload the logic into individual builders like: `buildPythonManylinuxPackage` or `buildPythonCondaPackage`.

FRidh commented:
> Why exactly do you want to get rid of the formats? If your motivation is to reduce complexity in mk-python-derivation.nix, we could offload the logic into individual builders like: buildPythonManylinuxPackage or buildPythonCondaPackage.

The hooks are useful also outside of buildPythonPackage. They were split out so you could have polyglot builders. In my opinion the format hides the hooks too much. I think if you would have to add them explicitly, like makeWrapper in other derivations, there would be more understanding.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 123323` at 8a80f95e run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>198 packages marked as broken and skipped:</summary>
  <ul>
    <li>bareos</li>
    <li>belle-sip</li>
    <li>blastem</li>
    <li>bonfire</li>
    <li>cassandra_2_1</li>
    <li>cassandra_2_2</li>
    <li>cura_stable</li>
    <li>cvc4</li>
    <li>deeptools</li>
    <li>displaycal</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>1959 packages skipped due to time constraints:</summary>
  <ul>
    <li>DisnixWebService</li>
    <li>MACS2</li>
    <li>R</li>
    <li>abcl</li>
    <li>adapta-gtk-theme</li>
    <li>adoptopenjdk-icedtea-web</li>
    <li>aiodnsbrute</li>
    <li>alloy (alloy5)</li>
    <li>alloy4</li>
    <li>almonds</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>326 packages built successfully:</summary>
  <ul>
    <li>actdiag (python38Packages.actdiag)</li>
    <li>black (python38Packages.black)</li>
    <li>blockdiag (python38Packages.blockdiag)</li>
    <li>botamusique</li>
    <li>corrscope</li>
    <li>dr14_tmeter</li>
    <li>giara</li>
    <li>image-analyzer</li>
    <li>imgp</li>
    <li>kissfft</li>
    <li>lndmanage</li>
    <li>markdown-pp</li>
    <li>nwdiag (python38Packages.nwdiag)</li>
    <li>pyspread</li>
    <li>python38Packages.Rtree</li>
    <li>python38Packages.aafigure</li>
    <li>python38Packages.accuweather</li>
    <li>python38Packages.aiobotocore</li>
    <li>python38Packages.aiohttp</li>
    <li>python38Packages.aiohttp-cors</li>
    <li>python38Packages.aiohttp-jinja2</li>
    <li>python38Packages.aiohttp-socks</li>
    <li>python38Packages.aiohttp-wsgi</li>
    <li>python38Packages.aiohue</li>
    <li>python38Packages.aionotion</li>
    <li>python38Packages.aiopvpc</li>
    <li>python38Packages.aioresponses</li>
    <li>python38Packages.airly</li>
    <li>python38Packages.aresponses</li>
    <li>python38Packages.arrayqueues</li>
    <li>python38Packages.asciimatics</li>
    <li>python38Packages.async-upnp-client</li>
    <li>python38Packages.autograd</li>
    <li>python38Packages.awkward0</li>
    <li>python38Packages.axis</li>
    <li>python38Packages.azure-core</li>
    <li>python38Packages.azure-keyvault-secrets</li>
    <li>python38Packages.azure-mgmt-common</li>
    <li>python38Packages.azure-mgmt-core</li>
    <li>python38Packages.azure-storage-blob</li>
    <li>python38Packages.backports_functools_lru_cache</li>
    <li>python38Packages.bokeh</li>
    <li>python38Packages.boto</li>
    <li>python38Packages.bottleneck</li>
    <li>python38Packages.brotlipy</li>
    <li>python38Packages.cairocffi</li>
    <li>python38Packages.cairosvg</li>
    <li>python38Packages.cgen</li>
    <li>python38Packages.cheroot</li>
    <li>python38Packages.clint</li>
    <li>python38Packages.colorcet</li>
    <li>python38Packages.colorspacious</li>
    <li>python38Packages.construct</li>
    <li>python38Packages.denonavr</li>
    <li>python38Packages.django-versatileimagefield</li>
    <li>python38Packages.dufte</li>
    <li>python38Packages.emcee</li>
    <li>python38Packages.faadelays</li>
    <li>python38Packages.fabulous</li>
    <li>python38Packages.face_recognition</li>
    <li>python38Packages.fastprogress</li>
    <li>python38Packages.folium</li>
    <li>python38Packages.freebox-api</li>
    <li>python38Packages.fsspec</li>
    <li>python38Packages.genpy</li>
    <li>python38Packages.geoalchemy2</li>
    <li>python38Packages.geoip2</li>
    <li>python38Packages.geojson-client</li>
    <li>python38Packages.google-api-core</li>
    <li>python38Packages.google-api-python-client</li>
    <li>python38Packages.google-auth</li>
    <li>python38Packages.google-auth-httplib2</li>
    <li>python38Packages.google-auth-oauthlib</li>
    <li>python38Packages.google-cloud-core</li>
    <li>python38Packages.google-cloud-iam</li>
    <li>python38Packages.google-cloud-kms</li>
    <li>python38Packages.google-cloud-storage</li>
    <li>python38Packages.google-cloud-testutils</li>
    <li>python38Packages.google-resumable-media</li>
    <li>python38Packages.handout</li>
    <li>python38Packages.haversine</li>
    <li>python38Packages.homematicip</li>
    <li>python38Packages.httpbin</li>
    <li>python38Packages.httpcore</li>
    <li>python38Packages.httplib2</li>
    <li>python38Packages.httpretty</li>
    <li>python38Packages.httpx</li>
    <li>python38Packages.hyperion-py</li>
    <li>python38Packages.imagecodecs-lite</li>
    <li>python38Packages.imageio</li>
    <li>python38Packages.iminuit</li>
    <li>python38Packages.jaraco_collections</li>
    <li>python38Packages.jaraco_functools</li>
    <li>python38Packages.jaraco_text</li>
    <li>python38Packages.javaobj-py3</li>
    <li>python38Packages.jsonrpc-async</li>
    <li>python38Packages.jsonrpc-websocket</li>
    <li>python38Packages.ledgerblue</li>
    <li>python38Packages.matplotlib</li>
    <li>python38Packages.matrix-nio</li>
    <li>python38Packages.mautrix (python38Packages.mautrix-appservice)</li>
    <li>python38Packages.mocket</li>
    <li>python38Packages.motioneye-client</li>
    <li>python38Packages.msgpack-numpy</li>
    <li>python38Packages.msrest</li>
    <li>python38Packages.msrestazure</li>
    <li>python38Packages.mutesync</li>
    <li>python38Packages.nose-randomly</li>
    <li>python38Packages.nose_progressive</li>
    <li>python38Packages.numba</li>
    <li>python38Packages.numexpr</li>
    <li>python38Packages.numpy</li>
    <li>python38Packages.oauth2client</li>
    <li>python38Packages.omnilogic</li>
    <li>python38Packages.openpaperwork-gtk</li>
    <li>python38Packages.phe</li>
    <li>python38Packages.piexif</li>
    <li>python38Packages.pilkit</li>
    <li>python38Packages.pillow</li>
    <li>python38Packages.pillowfight</li>
    <li>python38Packages.plexauth</li>
    <li>python38Packages.plexwebsocket</li>
    <li>python38Packages.plyfile</li>
    <li>python38Packages.pook</li>
    <li>python38Packages.portend</li>
    <li>python38Packages.pproxy</li>
    <li>python38Packages.pyclimacell</li>
    <li>python38Packages.pycollada</li>
    <li>python38Packages.pydeconz</li>
    <li>python38Packages.pyeconet</li>
    <li>python38Packages.pyerfa</li>
    <li>python38Packages.pyface</li>
    <li>python38Packages.pyhaversion</li>
    <li>python38Packages.pyipp</li>
    <li>python38Packages.pykeepass</li>
    <li>python38Packages.pykmtronic</li>
    <li>python38Packages.pykodi</li>
    <li>python38Packages.pylitterbot</li>
    <li>python38Packages.pyls-black</li>
    <li>python38Packages.pymazda</li>
    <li>python38Packages.pymbolic</li>
    <li>python38Packages.pymeteireann</li>
    <li>python38Packages.pymetno</li>
    <li>python38Packages.pyocr</li>
    <li>python38Packages.pyopengl</li>
    <li>python38Packages.pyphotonfile</li>
    <li>python38Packages.pypillowfight</li>
    <li>python38Packages.pyplaato</li>
    <li>python38Packages.pyrituals</li>
    <li>python38Packages.pyrr</li>
    <li>python38Packages.pysma</li>
    <li>python38Packages.pysmartthings</li>
    <li>python38Packages.soundfile (python38Packages.pysoundfile)</li>
    <li>python38Packages.pysrim</li>
    <li>python38Packages.pytesseract</li>
    <li>python38Packages.pytest-aiohttp</li>
    <li>python38Packages.pytest-arraydiff</li>
    <li>python38Packages.pytest-black</li>
    <li>python38Packages.pytest-doctestplus</li>
    <li>python38Packages.pytest-filter-subpackage</li>
    <li>python38Packages.pytest-httpbin</li>
    <li>python38Packages.pytest-httpx</li>
    <li>python38Packages.pytest-mpl</li>
    <li>python38Packages.pytest-randomly</li>
    <li>python38Packages.pytest-vcr</li>
    <li>python38Packages.python-engineio</li>
    <li>python38Packages.python-socketio</li>
    <li>python38Packages.pytools</li>
    <li>python38Packages.pyturbojpeg</li>
    <li>python38Packages.pywemo</li>
    <li>python38Packages.qimage2ndarray</li>
    <li>python38Packages.qrcode</li>
    <li>python38Packages.reportlab</li>
    <li>python38Packages.rokuecp</li>
    <li>python38Packages.rtmixer</li>
    <li>python38Packages.s3fs</li>
    <li>python38Packages.scapy</li>
    <li>python38Packages.seabreeze</li>
    <li>seqdiag (python38Packages.seqdiag)</li>
    <li>python38Packages.shapely</li>
    <li>python38Packages.sharedmem</li>
    <li>python38Packages.simplisafe-python</li>
    <li>python38Packages.slackclient</li>
    <li>python38Packages.smart_open</li>
    <li>python38Packages.smhi-pkg</li>
    <li>python38Packages.snapcast</li>
    <li>python38Packages.snuggs</li>
    <li>python38Packages.sockjs</li>
    <li>python38Packages.sounddevice</li>
    <li>python38Packages.spglib</li>
    <li>python38Packages.sphfile</li>
    <li>python38Packages.tables</li>
    <li>python38Packages.tempora</li>
    <li>python38Packages.tifffile</li>
    <li>python38Packages.traits</li>
    <li>trezorctl (python38Packages.trezor)</li>
    <li>python38Packages.trimesh</li>
    <li>python38Packages.twofish</li>
    <li>python38Packages.uncertainties</li>
    <li>python38Packages.uproot3-methods</li>
    <li>python38Packages.uvicorn</li>
    <li>python38Packages.uvloop</li>
    <li>python38Packages.vcrpy</li>
    <li>python38Packages.websockify</li>
    <li>python38Packages.willow</li>
    <li>xdot (python38Packages.xdot)</li>
    <li>python38Packages.xpybutil</li>
    <li>python38Packages.zeep</li>
    <li>python38Packages.zwave-js-server-python</li>
    <li>python39Packages.Rtree</li>
    <li>python39Packages.aafigure</li>
    <li>python39Packages.aiobotocore</li>
    <li>python39Packages.aiohttp</li>
    <li>python39Packages.aiohttp-cors</li>
    <li>python39Packages.aioresponses</li>
    <li>python39Packages.aresponses</li>
    <li>python39Packages.arrayqueues</li>
    <li>python39Packages.asciimatics</li>
    <li>python39Packages.autograd</li>
    <li>python39Packages.azure-core</li>
    <li>python39Packages.azure-storage-blob</li>
    <li>python39Packages.backports_functools_lru_cache</li>
    <li>python39Packages.black</li>
    <li>python39Packages.blockdiag</li>
    <li>python39Packages.bokeh</li>
    <li>python39Packages.bottleneck</li>
    <li>python39Packages.brotlipy</li>
    <li>python39Packages.cairocffi</li>
    <li>python39Packages.cairosvg</li>
    <li>python39Packages.cgen</li>
    <li>python39Packages.clint</li>
    <li>python39Packages.colorcet</li>
    <li>python39Packages.construct</li>
    <li>python39Packages.dufte</li>
    <li>python39Packages.emcee</li>
    <li>python39Packages.fabulous</li>
    <li>python39Packages.face_recognition</li>
    <li>python39Packages.fastprogress</li>
    <li>python39Packages.fsspec</li>
    <li>python39Packages.genpy</li>
    <li>python39Packages.geoalchemy2</li>
    <li>python39Packages.geoip2</li>
    <li>python39Packages.google-api-core</li>
    <li>python39Packages.google-api-python-client</li>
    <li>python39Packages.google-auth</li>
    <li>python39Packages.google-auth-httplib2</li>
    <li>python39Packages.google-auth-oauthlib</li>
    <li>python39Packages.haversine</li>
    <li>python39Packages.httpbin</li>
    <li>python39Packages.httpcore</li>
    <li>python39Packages.httplib2</li>
    <li>python39Packages.httpretty</li>
    <li>python39Packages.httpx</li>
    <li>python39Packages.imagecodecs-lite</li>
    <li>python39Packages.jaraco_collections</li>
    <li>python39Packages.jaraco_functools</li>
    <li>python39Packages.jaraco_text</li>
    <li>python39Packages.javaobj-py3</li>
    <li>python39Packages.ledgerblue</li>
    <li>python39Packages.matplotlib</li>
    <li>python39Packages.memory_profiler</li>
    <li>python39Packages.mocket</li>
    <li>python39Packages.moderngl-window</li>
    <li>python39Packages.msgpack-numpy</li>
    <li>python39Packages.msrest</li>
    <li>python39Packages.nose-randomly</li>
    <li>python39Packages.nose_progressive</li>
    <li>python39Packages.numexpr</li>
    <li>python39Packages.numpy</li>
    <li>python39Packages.oauth2client</li>
    <li>python39Packages.phe</li>
    <li>python39Packages.piexif</li>
    <li>python39Packages.pilkit</li>
    <li>python39Packages.pillow</li>
    <li>python39Packages.pillowfight</li>
    <li>python39Packages.plyfile</li>
    <li>python39Packages.pook</li>
    <li>python39Packages.portend</li>
    <li>python39Packages.pproxy</li>
    <li>python39Packages.pydenticon</li>
    <li>python39Packages.pyerfa</li>
    <li>python39Packages.pyface</li>
    <li>python39Packages.pykeepass</li>
    <li>python39Packages.pymbolic</li>
    <li>python39Packages.pyocr</li>
    <li>python39Packages.pypillowfight</li>
    <li>python39Packages.pyrr</li>
    <li>python39Packages.soundfile (python39Packages.pysoundfile)</li>
    <li>python39Packages.pysrim</li>
    <li>python39Packages.pytesseract</li>
    <li>python39Packages.pytest-aiohttp</li>
    <li>python39Packages.pytest-arraydiff</li>
    <li>python39Packages.pytest-black</li>
    <li>python39Packages.pytest-doctestplus</li>
    <li>python39Packages.pytest-filter-subpackage</li>
    <li>python39Packages.pytest-httpbin</li>
    <li>python39Packages.pytest-mpl</li>
    <li>python39Packages.pytest-randomly</li>
    <li>python39Packages.pytest-vcr</li>
    <li>python39Packages.pytools</li>
    <li>python39Packages.qimage2ndarray</li>
    <li>python39Packages.quantities</li>
    <li>python39Packages.reportlab</li>
    <li>python39Packages.s3fs</li>
    <li>python39Packages.seabreeze</li>
    <li>python39Packages.shapely</li>
    <li>python39Packages.sharedmem</li>
    <li>python39Packages.snapcast</li>
    <li>python39Packages.snuggs</li>
    <li>python39Packages.sounddevice</li>
    <li>python39Packages.spglib</li>
    <li>python39Packages.sphfile</li>
    <li>python39Packages.tables</li>
    <li>python39Packages.tempora</li>
    <li>python39Packages.tifffile</li>
    <li>python39Packages.traits</li>
    <li>python39Packages.twofish</li>
    <li>python39Packages.ueberzug</li>
    <li>python39Packages.uncertainties</li>
    <li>python39Packages.uvicorn</li>
    <li>python39Packages.uvloop</li>
    <li>python39Packages.vcrpy</li>
    <li>python39Packages.willow</li>
    <li>python39Packages.xdot</li>
    <li>sl1-to-photon</li>
    <li>vcsi</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `setuptoolsBuildHook`.
Near pkgs/development/python-modules/numpy/default.nix:13:3:
```
   |
13 | , setuptoolsBuildHook
   |   ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/interpreters/python/hooks/conda-install-hook.sh:
```diff
@@ -0,0 +1,32 @@
+
+    # remove condas place holder for the interpreter path.
+    if [ -e $out/bin ]; then
+      find $out/bin -type f -exec sed -i "s|/opt/anaconda1anaconda2anaconda3||g" {} \;
```
You are sure you are not missing a separator here?

DavHau replied:
Separator at what place? It is a while ago when I wrote this. I will try to find a conda package which is affected by this and test it again. The reason I had to remove the `/opt/anaconda1anaconda2anaconda3` prefix from shebangs might be because `patchPythonScript`wouldn't work correctly otherwise. Does that make sense? Anyways, I will check it again.

DavHau replied:
Some python scripts in conda packages start with this line:
`#!/opt/anaconda1anaconda2anaconda3/bin/python`

DavHau replied:
I removed the replacement now. I tested it and shebangs were patched correctly anyways.

SuperSandro2000 replied:
> Separator at what place?

``/opt/anaconda1anaconda2anaconda3`` looks very strange. I thought it would need to be ``/opt/anaconda1,/opt/anaconda2,/opt/anaconda3``

SuperSandro2000 commented on pkgs/development/interpreters/python/hooks/conda-unpack-hook.sh:
```diff
@@ -0,0 +1,18 @@
+    # use lbzip2 for parallel decompression (bz2 is slow)
+    lbzip2 -dc -n $NIX_BUILD_CORES $src | tar --exclude='info' -x
+
+#     runHook postUnpack # Calls find...?
```
```suggestion
    # runHook postUnpack # Calls find...?
```

DavHau replied:
I copied this line from another python hook. I'd really like to get rid of this completely. I think we should not disable postUnpack but instead fix the problem it is causing. But If I fix this here, it will cause a mass rebuild, so I better keep this for another PR. Though, I will fix the indentation as you proposed.

SuperSandro2000 commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -143,7 +166,12 @@ let
     ] ++ nativeBuildInputs;
 
-    buildInputs = buildInputs ++ pythonPath;
+    buildInputs =
+      buildInputs ++ pythonPath
+      ++ lib.optionals (format == "condabin") (
+        [ alsaLib cups gcc-unwrapped.lib libGL ]
+        ++ (with xorg; [ libSM libICE libX11 libXau libXdamage libXi libXrender libXrandr libXcomposite libXcursor libXtst libXScrnSaver])
```
```suggestion
    buildInputs = buildInputs ++ pythonPath
      ++ lib.optionals (format == "condabin") ([ alsaLib cups gcc-unwrapped.lib libGL ]
        ++ (with xorg; [ libSM libICE libX11 libXau libXdamage libXi libXrender libXrandr libXcomposite libXcursor libXtst libXScrnSaver])
```

DavHau commented:
> We also need to agree on who will maintain additions like these in Nixpkgs.
I'm happy to maintain the conda related hooks/packages. Is there any way I can add myself as a maintainer? Hooks don't have a meta field as far as I can tell.

DavHau commented:
I reverted all changes to `mk-python-deriaviton.nix` and instead included a `pkgs.pythonCondaPackages.condaPatchelfLibs`.

FRidh commented:
Note I don't want to dictate how things should be in Nixpkgs/Python. We're a community here, and you are contributing a lot to the ecosystem. I also have hardly time left for Nixpkgs. The thing is, everything we add can be used and then you end up "supporting" it. The question is what we want to support!

> In general I'm not a fan of shifting more work to the user. I also think that installing manylinux wheels is still too complicated.
I even would have proposed to create a format manylinux that has everything on board.

Yes, I understand your view. I still have a bit the hope that someday we would not need to have custom builders, and be able to do everything with hooks (although there are clear downsides to that as well.

As it is, this change is good to me. About adding the `format` change (also for manylinux), I suggest opening a different issue/PR. I am sure others would be interested in it.

> I'm happy to maintain the conda related hooks/packages. Is there any way I can add myself as a maintainer? Hooks don't have a meta field as far as I can tell.

Great. One option for things other than derivations is to use the `CODEOWNERS` file.

FRidh commented:
Note you may also want to add a test in `pkgs/development/interpreters/python/tests.nix`, especially because we have no package actually using this in Nixpkgs.

DavHau commented:
> The thing is, everything we add can be used and then you end up "supporting" it. The question is what we want to support!

This is a very good point I haven't yet thought about very much. It is probably better to first just bring the feature without a finished UI, then see which patterns evolve in the community and later when it's more clear how things are getting used, maybe add some more simplification to it.

Thanks, I haven't known about the `CODEOWNERS`. I will add myself there and also add some test for this.

jonringer reviewed (commented):
Unless I missed something, these hooks like they will be ran unconditionally. Should we have some test to see if the package in question is a conda package?

DavHau commented:
I now added some test to `pkgs/development/interpreters/python/tests.nix` and added myself to `CODEOWNERS`

>Unless I missed something, these hooks like they will be ran unconditionally. Should we have some test to see if the package in question is a conda package?

Not sure what you mean be `unconditionally`. The hooks are not included by default. Only if you manually add them to `nativeBuildInputs` they will be executed. Do you think it still needs an extra check? The other hooks don't have such checks AFAIK.

FRidh commented on .github/CODEOWNERS:
```diff
@@ -71,12 +71,14 @@
+/pkgs/development/python-modules                            @FRidh @jonringer
+/doc/languages-frameworks/python.section.md                 @FRidh
+/pkgs/development/tools/poetry2nix                          @adisbladis
+/pkgs/development/interpreters/python/hooks                 @DavHau
```
Could you prefix me and @jonringer for the hooks folder

DavHau replied:
done

FRidh commented on pkgs/development/interpreters/python/tests.nix:
```diff
@@ -121,4 +121,38 @@ let
+    requests = callPackage ({
+        autoPatchelfHook,
+        fetchurl,
+        python38,
```
The tests are executed for each Python interpreter. Therefore, here you should pass `python` in, and then disable the test for Python versions not supported. See a bit above where Python versions are limited as well.

DavHau replied:
Thanks. I now modified the test to work for all interpreter versions.

FRidh commented on pkgs/development/interpreters/python/tests.nix:
```diff
@@ -121,4 +121,37 @@ let
+    in
+    {
+      condaExamplePackage = runCommand "import-requests" {} ''
+        ${pythonWithRequests}/bin/python -c "import requests" > $out
```
`pythonWithRequests.interpreter` so it could also work with PyPy.

DavHau replied:
done
